File: 131PalindromePartitioning/PalindromePartitioning.cpp


#include <vector>
#include <string>
#include <cstdlib>

class PalindromePartitioning {
public:
    std::vector<std::vector<std::string> > partition3(std::string s) {
        std::vector<std::vector<std::string> > ans;
        std::vector<std::string> cur;
        partitionHelper(s, ans, cur, 0);
        return ans;
    }

private:
    void partitionHelper(
        std::string& s,
        std::vector<std::vector<std::string> >& ans,
        std::vector<std::string>& cur,
        size_t idx
    ) {
        if (idx >= s.size()) {
            // all letters checked
            ans.push_back(cur);
            return;
        }
        for (size_t i = idx; i < s.size(); ++i) {
            if (isPalindrome(s.substr(idx, i - idx + 1))) {
                cur.push_back(s.substr(idx, i - idx + 1));
                partitionHelper(s, ans, cur, i + 1);
                cur.pop_back();
            }
        }
    }

    bool isPalindrome(std::string s) {
        size_t l, r;
        l = 0;
        r = s.size() - 1;
        while (l < r) {
            if (s[l++] != s[r--]) return false;
        }
        return true;
    }
};


int main() {
    return EXIT_SUCCESS;
}

File: 0131PalindromePartitioning/PalindromePartitioning.py
# -*- coding: utf-8 -*-

from typing import List

class PalindromePartitioning:
    def partition3(self, s: "str") -> "List[List[str]]":
        ans = []
        self.__partitionHelper(s, ans, [], 0)
        return ans

    def __partitionHelper(
        self, s: "str", ans: "List[List[str]]", cur: "List[str]", idx: "int"
    ) -> "None":
        if idx >= len(s):
            ans.append(cur[:])
            return
        for i in range(idx, len(s)):
            if self.__isPalindrome(s[idx : i + 1]):
                cur.append(s[idx : i + 1])
                self.__partitionHelper(s, ans, cur, i + 1)
                ## start from i+1 for next word
                cur.pop()

    def __isPalindrome(self, s: "str") -> "bool":
        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]: return False
            l += 1
            r -= 1
        return True

def main():
    test = PalindromePartitioning()
    print(test.partition3("aab"))

if __name__ == "__main__":
    main()
